### PR TITLE
Refactoring model structure

### LINKED ISSUE #4: Refactoring model structure

### PR DESCRIPTION
**Pull request by pro100user15**
- Create entity Address and Publishing
- Remove entity PublishingHouse, WorldLiterature and PeriodLiterature
- Status entity rewritten to Status enum

(And create base CRUD operation for new entities)
closes #4

### CODE REVIEW

File: bookstore-backend/src/main/java/com/pro100user/bookstore/model/Book.java [hunk 1]
```diff
@@ -72,12 +71,6 @@
     @NotEmpty(message = "Publication year cannot be empty")
     private int yearPublication;
 
-    @NotNull
-    @Column(nullable = false)
-    @Max(value = 2022, message = "First edition year cannot be higher than 2022")
```
Comment (RuslanSlobodian) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/model/Book.java:77]: current year is used. what will be next year?

**Review by RuslanSlobodian — CHANGES_REQUESTED** [review]

File: bookstore-backend/src/main/java/com/pro100user/bookstore/model/Basket.java [hunk 2]
```diff
@@ -39,5 +39,5 @@
     @Column(nullable = false)
     @Min(value = 0, message = "Price cannot be less than 0")
     @NotEmpty(message = "Price cannot be empty")
-    private double price;
+    private double totalPrice;
```
Comment (RuslanSlobodian) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/model/Basket.java:42]: Price is better to save as a BigDecimal value

File: bookstore-backend/src/main/java/com/pro100user/bookstore/model/Book.java (hunk 1 above)
Comment (RuslanSlobodian) [reply bookstore-backend/src/main/java/com/pro100user/bookstore/model/Book.java:77]: if you use java.time.Year, you may set annotation @PastOrPresent

File: bookstore-backend/src/main/java/com/pro100user/bookstore/model/Address.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    @Size(min = 1, max = 10, message = "Number must be between 1 and 10 characters long")
+    @Column(length = 10, nullable = false)
+    @NotEmpty(message = "Number cannot be empty")
+    public String number;
+
+    @OneToOne(mappedBy = "address")
+    Publishing publishing;
+
+    @OneToMany(mappedBy = "address")
+    private List<User> users;
```
Comment (RuslanSlobodian) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/model/Address.java:52]: Setters for fields with annotation @ManyToMany and @OneToMany must be private:
Setter(AccessLevel.PRIVATE);

**Review by RuslanSlobodian — CHANGES_REQUESTED** [review]
Setters for fields with annotation @ManyToMany and @OneToMany must be private:
Setter(AccessLevel.PRIVATE);

File: bookstore-backend/src/main/java/com/pro100user/bookstore/model/Publishing.java [hunk 4]
```diff
@@ -12,20 +14,24 @@
...
+@Table(name = "publishing")
+public class Publishing implements Serializable {
 
     @Id
     @GeneratedValue(strategy = GenerationType.IDENTITY)
-    @Setter(AccessLevel.NONE)
     private Long id;
 
     @NotNull
     @Size(min = 1, max = 64, message = "Publishing house must be between 1 and 64 characters long")
```
Comment (dshma) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/model/Publishing.java:25]: still house

File: bookstore-backend/src/main/java/com/pro100user/bookstore/model/Publishing.java [hunk 5]
```diff
@@ -12,20 +14,24 @@
...
     @Column(length = 64, nullable = false)
     @NotEmpty(message = "Publishing house cannot be empty")
```
Comment (dshma) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/model/Publishing.java:27]: still house

File: bookstore-backend/src/main/java/com/pro100user/bookstore/model/Publishing.java (hunk 4 above)
Comment (dshma) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/model/Publishing.java:25]: I don't think 64 characters is enough? Don't be greedy :)

File: bookstore-backend/src/main/java/com/pro100user/bookstore/model/Address.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    @NotEmpty(message = "Street cannot be empty")
+    public String street;
+
+    @NotNull
+    @Size(min = 1, max = 10, message = "Number must be between 1 and 10 characters long")
+    @Column(length = 10, nullable = false)
+    @NotEmpty(message = "Number cannot be empty")
+    public String number;
+
+    @OneToOne(mappedBy = "address")
```
Comment (dshma) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/model/Address.java:48]: avoid EAGER fetch type

File: bookstore-backend/src/main/java/com/pro100user/bookstore/model/Publishing.java [hunk 7]
```diff
@@ -12,20 +14,24 @@
...
-    private String name;
+    private String publishing_name;
 
-    @OneToMany(mappedBy = "publishingHouse")
+    @NotNull
+    @OneToOne(targetEntity = Address.class)
```
Comment (dshma) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/model/Publishing.java:31]: avoid EAGER fetch type

File: bookstore-backend/src/main/java/com/pro100user/bookstore/model/User.java [hunk 8]
```diff
@@ -62,13 +62,12 @@
...
-    @Size(min = 4, max = 64, message = "Address must be between 4 and 64 characters long")
-    @NotEmpty(message = "Address cannot be empty")
-    private String address;
+    @ManyToOne(fetch = FetchType.LAZY, targetEntity = Address.class)
+    @JoinColumn(name = "address_id")
+    private Address address;
 
     @NotNull
-    private LocalDateTime createdAt = LocalDateTime.now();
+    private LocalDateTime createdAt;
```
Comment (dshma) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/model/User.java:65]: this column cannot be null, updatable and available for creation

File: bookstore-backend/src/main/java/com/pro100user/bookstore/repository/impl/AddressRepositoryImpl.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+import com.pro100user.bookstore.model.Address;
+import com.pro100user.bookstore.repository.AddressRepository;
+import org.hibernate.SessionFactory;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public class AddressRepositoryImpl extends BasicRepositoryImpl<Address, Long> implements AddressRepository {
+
+    public AddressRepositoryImpl(SessionFactory sessionFactory) {
```
Comment (dshma) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/repository/impl/AddressRepositoryImpl.java:11]: BasicRepositoryImpl already autowires this bean

File: bookstore-backend/src/main/java/com/pro100user/bookstore/repository/impl/PublishingRepositoryImpl.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+import com.pro100user.bookstore.model.Publishing;
+import com.pro100user.bookstore.repository.PublishingRepository;
+import org.hibernate.SessionFactory;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public class PublishingRepositoryImpl extends BasicRepositoryImpl<Publishing, Long> implements PublishingRepository {
+
+    public PublishingRepositoryImpl(SessionFactory sessionFactory) {
```
Comment (dshma) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/repository/impl/PublishingRepositoryImpl.java:11]: BasicRepositoryImpl already autowires this bean

File: bookstore-backend/src/main/java/com/pro100user/bookstore/service/impl/AddressServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    public AddressServiceImpl(AddressRepository addressRepository) {
+        this.addressRepository = addressRepository;
+    }
+
+    @Override
+    public Address create(Address object) {
+        return addressRepository.create(object);
+    }
+
+    @Transactional(readOnly = true)
```
Comment (dshma) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/service/impl/AddressServiceImpl.java:26]: do you need a transaction here?

File: bookstore-backend/src/main/java/com/pro100user/bookstore/service/impl/AddressServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+public class AddressServiceImpl implements AddressService {
+
+    private final AddressRepository addressRepository;
+
+    public AddressServiceImpl(AddressRepository addressRepository) {
+        this.addressRepository = addressRepository;
+    }
+
+    @Override
+    public Address create(Address object) {
```
Comment (dshma) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/service/impl/AddressServiceImpl.java:22]: what is the purpose of the BasicService if you implement all CRUD methods in all impl services?

File: bookstore-backend/src/main/java/com/pro100user/bookstore/service/impl/AddressServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+    @Transactional(readOnly = true)
+    @Override
+    public Address readById(Long id) {
+        return addressRepository.readById(id);
+    }
+
+    @Override
+    public Address update(Address object) {
+        readById(object.getId());
```
Comment (dshma) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/service/impl/AddressServiceImpl.java:34]: validation should be extracted into a separate method

**Review by dshma — CHANGES_REQUESTED** [review]

File: bookstore-backend/src/main/java/com/pro100user/bookstore/model/User.java [hunk 14]
```diff
@@ -22,18 +23,15 @@
 
     @Id
     @GeneratedValue(strategy = GenerationType.IDENTITY)
-    @Setter(AccessLevel.NONE)
     private Long id;
 
-    @NotNull
     @Size(min = 2, max = 35, message = "Name must be between 2 and 35 characters long")
```
Comment (dshma) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/model/User.java:28]: why 35 characters?)

File: bookstore-backend/src/main/java/com/pro100user/bookstore/repository/CategoryRepository.java [hunk 15]
```diff
@@ -2,13 +2,12 @@
...
-import com.pro100user.bookstore.model.model.CategoryModel;
 
 import java.util.List;
 
 public interface CategoryRepository extends BasicRepository<Category, Long> {
 
     List<Category> findByName(String name);
     List<Book> getListBookByCategoryName(String name);
-    List<CategoryModel> getCategoryModels();
+    Integer getBookSizeByCategoryName(Long id);
```
Comment (dshma) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/repository/CategoryRepository.java:12]: ByCategoryName but the parameter is ID

File: bookstore-backend/src/main/java/com/pro100user/bookstore/repository/impl/CategoryRepositoryImpl.java [hunk 16]
```diff
@@ -16,8 +14,9 @@
 public class CategoryRepositoryImpl extends BasicRepositoryImpl<Category, Long> implements CategoryRepository {
 
     private String SELECT_BY_NAME = "FROM Category WHERE name like :name";
+    private String SELECT_COUNT_BOOKS = "SELECT books.size FROM Category WHERE id=:id";
     private String SELECT_BOOKS_BY_CATEGORY_NAME = "FROM Category.books WHERE Category.name like :name";
-    private String SELECT_CATEGORY_WITH_COUNT_BOOKS = "SELECT new com.pro100user.bookstore.entity.model.CategoryModel(id, name, books.size) from Category";
+    //private String SELECT_CATEGORY_WITH_COUNT_BOOKS = "SELECT new com.pro100user.bookstore.entity.model.CategoryModel(id, name, books.size) from Category";
```
Comment (dshma) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/repository/impl/CategoryRepositoryImpl.java:19]: avoid commented code

File: bookstore-backend/src/main/java/com/pro100user/bookstore/repository/impl/CategoryRepositoryImpl.java [hunk 17]
```diff
@@ -35,11 +34,16 @@ public CategoryRepositoryImpl(SessionFactory sessionFactory) {
                         .setParameter("name", name).getResultList();
     }
 
-    @Override
+    public Integer getBookSizeByCategoryName(Long id) {
+        return (Integer) sessionFactory.getCurrentSession().createQuery(SELECT_COUNT_BOOKS)
```
Comment (dshma) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/repository/impl/CategoryRepositoryImpl.java:38]: wouldn't int be enough?

File: bookstore-backend/src/main/java/com/pro100user/bookstore/repository/impl/CategoryRepositoryImpl.java [hunk 18]
```diff
@@ -35,11 +34,16 @@ public CategoryRepositoryImpl(SessionFactory sessionFactory) {
...
+                .setParameter("id", id).getSingleResult();
+    }
+
+    /*@Override
```
Comment (dshma) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/repository/impl/CategoryRepositoryImpl.java:42]: avoid commented code

File: bookstore-backend/src/main/java/com/pro100user/bookstore/repository/impl/CategoryRepositoryImpl.java [hunk 19]
```diff
@@ -16,8 +14,9 @@
 public class CategoryRepositoryImpl extends BasicRepositoryImpl<Category, Long> implements CategoryRepository {
 
     private String SELECT_BY_NAME = "FROM Category WHERE name like :name";
```
Comment (dshma) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/repository/impl/CategoryRepositoryImpl.java:16]: these class variables should be static final

File: bookstore-backend/src/main/java/com/pro100user/bookstore/repository/impl/UserRepositoryImpl.java [hunk 20]
```diff
@@ -19,8 +21,11 @@ public UserRepositoryImpl(SessionFactory sessionFactory) {
 
     @Override
     public User findByLogin(String login) {
-        List<User> users = sessionFactory.getCurrentSession().createQuery(findUserByLogin, User.class)
-                .setParameter("login", login).getResultList();
-        return users.isEmpty() ? null : users.get(0);
+        User user = sessionFactory.getCurrentSession().createQuery(findUserByLogin, User.class)
```
Comment (dshma) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/repository/impl/UserRepositoryImpl.java:24]: findUserByLogin should follow the constant's modifiers and naming convention

File: bookstore-backend/src/main/java/com/pro100user/bookstore/service/BookService.java [hunk 21]
```diff
@@ -2,5 +2,10 @@
 
 import com.pro100user.bookstore.model.Book;
 
+import java.util.List;
+
 public interface BookService extends BasicService<Book, Long> {
+
+    //TODO getListBookByCategoryName(String name);
```
Comment (dshma) [inline bookstore-backend/src/main/java/com/pro100user/bookstore/service/BookService.java:9]: avoid committing such code

File: bookstore-frontend/src/router/index.ts [hunk 22]
```diff
@@ -23,33 +24,37 @@ export enum RouteNames {
...
 
-        if(loading === false)
-            setLoading(true);
-    }, [roles, loading]);
+        if(roles && roles.includes(Role.MODERATOR)) {
+            const guestRoutes: IRote[] = [
+                {path: RouteNames.CATEGORIES, component: Category}
+            ];
+
+            console.log(guestRoutes);
```
Comment (dshma) [inline bookstore-frontend/src/router/index.ts:53]: avoid console logs

**Review by dshma — CHANGES_REQUESTED** [review]

**Review by dshma — APPROVED** [review]

File: bookstore-frontend/src/router/index.ts (hunk 22 above)
Comment (pro100user15) [reply bookstore-frontend/src/router/index.ts:53]: I used them to derive intermediate results. I'll take them

File: bookstore-backend/src/main/java/com/pro100user/bookstore/model/Address.java [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+    Publishing publishing;
```
[the remark this answers is not in the document]
Comment (pro100user15) [reply bookstore-backend/src/main/java/com/pro100user/bookstore/model/Address.java:49]: sorry I didn't notice

**Pull request merged**